src/box/gc.c

```diff
@@ -820,6 +861,449 @@ gc_consumer_update(const struct tt_uuid *uuid, const struct vclock *vclock)
+	key_end = mp_encode_str0(key_end, tt_uuid_str(uuid));
+	assert((unsigned long)(key_end - key_buf) < sizeof(key_buf));
+
+	char ops_buf[VCLOCK_STR_LEN_MAX + 20];
+	char *ops_end = ops_buf;
+	ops_end = mp_encode_array(ops_end, 1);
+	ops_end = mp_encode_array(ops_end, 3);
+	ops_end = mp_encode_str0(ops_end, "=");
+	ops_end = mp_encode_uint(ops_end, 1);
+	ops_end = mp_encode_vclock_ignore0(ops_end, vclock);
+	assert((unsigned long)(ops_end - ops_buf) < sizeof(ops_buf));
+
+	return box_update(BOX_GC_CONSUMERS_ID, 0, key_buf, key_end,
+			  ops_buf, ops_end, 0, NULL);
+}
+
+int
+gc_consumer_update(const struct tt_uuid *uuid, const struct vclock *vclock)
+{
+	assert(in_txn() == NULL);
+	gc_consumer_discard_async_updates(uuid);
+	if (!gc_consumer_is_persistent()) {
+		struct gc_consumer *consumer = gc_consumer_by_uuid(uuid);
+		if (consumer != NULL)
+			gc_consumer_update_impl(consumer, vclock);
+		return 0;
+	}
+
+	struct txn *txn = txn_begin();
+	if (txn == NULL)
+		return -1;
+	txn_set_flags(txn, TXN_FORCE_ASYNC);
+
+	int rc = gc_consumer_persistent_update_impl(uuid, vclock);
+
+	if (rc == 0)
+		rc = txn_commit(txn);
+	else
+		txn_abort(txn);
+	return rc;
+}
+
+bool
+gc_consumer_is_persistent(void)
+{
+	return space_by_id(BOX_GC_CONSUMERS_ID) != NULL;
+}
+
+void
+gc_consumer_update_async(const struct tt_uuid *uuid,
+			 const struct vclock *vclock)
+{
+	struct gc_consumer *consumer = gc_consumer_by_uuid(uuid);
+	if (consumer == NULL)
+		return;
+
+	if (!gc_consumer_is_persistent()) {
+		gc_consumer_update_impl(consumer, vclock);
+		return;
+	}
+
+	vclock_copy(&consumer->volatile_vclock, vclock);
+	consumer->is_async_updated = true;
+	/*
+	 * Do not wake up the background fiber if it's persisting
+	 * consumers right now because it may be waiting for a cbus
+	 * message, which doesn't tolerate spurious wakeups.
+	 */
+	if (gc.persist_scheduled++ == gc.persist_completed)
+		fiber_wakeup(gc.persist_fiber);
+}
+
+static int
+gc_consumers_persist_all(void)
+{
+	ERROR_INJECT(ERRINJ_WAL_GC_PERSIST_FIBER, {
+		diag_set(ClientError, ER_INJECTION, "WAL GC persist fiber");
+		return -1;
+	});
+	int rc = 0;
+	struct txn *txn = txn_begin();
+	if (txn == NULL)
+		return -1;
+	/* No need to wait limbo, so let's set TXN_FORCE_ASYNC here. */
+	txn_set_flags(txn, TXN_FORCE_ASYNC);
+
+	gc_consumer_hash_foreach(&gc.consumers_hash, consumer) {
+		struct vclock *old_vclock = &consumer->vclock;
+		struct vclock *new_vclock = &consumer->volatile_vclock;
+		struct tt_uuid *uuid = &consumer->uuid;
+
+		if (consumer->is_async_updated &&
+		    vclock_compare(old_vclock, new_vclock) != 0 &&
+		    gc_consumer_persistent_update_impl(uuid, new_vclock) != 0)
+			goto fail;
+	}
+	rc = txn_commit(txn);
+	goto out;
+fail:
+	rc = -1;
+	txn_abort(txn);
+out:
+	return rc;
+}
+
+/* Retry updating persistent WAL GC state every N seconds. */
+enum {
+#ifndef NDEBUG
+	WAL_GC_PERSIST_FIBER_RETRY_TIMEOUT = 1,
+#else
+	WAL_GC_PERSIST_FIBER_RETRY_TIMEOUT = 10,
```

Comment:
Why different timeouts for debug and non-debug?

Reply:
I think retry timeout should be quite large - no point trying to write to database too often if we fail to do it. On the other hand, when I test retries with `errinj`, I don't want to wait for this large timeout - that's why I have a small one for Debug build.
By the way, I'm [not the first](https://github.com/tarantool/tarantool/blob/c06d0d1491572a489f07bce9a355e2a95608bc7e/src/box/memtx_space.c#L61) one who did it.

Reply:
Covered this place with comment, reformatted a bit along the way.

Reply:
> By the way, I'm [not the first](https://github.com/tarantool/tarantool/blob/c06d0d1491572a489f07bce9a355e2a95608bc7e/src/box/memtx_space.c#L61) one who did it.

Then I'm ok with such approach. It just seemed to me, that we deal with such things by introducing errinj